### PR TITLE
update approvalYears.py

### PR DESCRIPTION
**Pull request by southalln**
Update it to python3!
And add CBER ingredients approval dates from purplebook

### CODE REVIEW

File: scripts/approvalYears.py [hunk 1]
```diff
@@ -148,7 +151,7 @@ def parseIngred(ing, uniiPT, uniiALL, missing):
                 if len(match) == 10:
                     ingreds.append(match)
                 else:
-                    if not missing.has_key(si):
+                    if si not in missing:
```
Comment (grishagin) [inline scripts/approvalYears.py:291]: Typo: "si"

File: scripts/approvalYears.py [hunk 2]
```diff
@@ -70,7 +73,7 @@ def getMainDir():
     return curr
 
 def resolveName(name): # returns empty string if resolver can't do anything with this name
-    if name in resolverCache.keys():
+    if name in list(resolverCache.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:210]: Should just be "if name in resolverCache:" (unless I'm missing something)

File: scripts/approvalYears.py [hunk 3]
```diff
@@ -198,7 +201,7 @@ def apprDateRegression(prods):
     window = 10000
     dataxs = dict()
     datays = dict()
-    for prod in prods.keys():
+    for prod in list(prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:361]: Same as above: "for prod in prods:"

File: scripts/approvalYears.py [hunk 4]
```diff
@@ -222,15 +225,15 @@ def apprDateRegression(prods):
                 datays[kind2].append(ts)
 
     models = dict()
-    for kind in dataxs.keys():
+    for kind in list(dataxs.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:385]: Same as above: "for kind in dataxs:"

File: scripts/approvalYears.py [hunk 5]
```diff
@@ -222,15 +225,15 @@ def apprDateRegression(prods):
...
         #print kind, len(dataxs[kind])
         if len(dataxs[kind]) > 20:
             regr = numpy.poly1d(numpy.polyfit(dataxs[kind], datays[kind], 1))
             models[kind] = regr
     figs = ['NDA', 'BLA', 'ANDA']
     for figt in figs:
         fig, ax = plt.subplots()
-        for kind in models.keys():
+        for kind in list(models.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:393]: Same as above.

File: scripts/approvalYears.py [hunk 6]
```diff
@@ -243,14 +246,14 @@ def apprDateRegression(prods):
         fig.tight_layout()
         plt.savefig(figt+'.png')
 
-    for prod in prods.keys():
+    for prod in list(prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:406]: Same as above.

File: scripts/approvalYears.py [hunk 7]
```diff
@@ -318,40 +321,40 @@ def writeInitApp(outfp, unii, early, earlyDate, myunii):
     #comment = apptype+'|'+appno+'|'+appsponsor+'|'+product+'|'+appurl+'|'
     #comment = comment + early[0].decode('latin-1').encode('ascii', 'replace')
     #outline = unii+"\tApproval Year\t"+year+"\t\t"+comment+"\t"+date+"\t"+method+"\n"
-    comment = early[0].decode('latin-1').encode('ascii', 'replace')
+    comment = early[0]# .decode('latin-1').encode('ascii', 'replace')
```
Comment (grishagin) [inline scripts/approvalYears.py:481]: Out of curiosity: why is this not needed anymore?

File: scripts/approvalYears.py [hunk 8]
```diff
@@ -549,7 +552,7 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
     fp = open(missingUNIIsfile, 'w')
     m2 = []
 
-    for key in missing.keys():
+    for key in list(missing.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:800]: Same as above.

File: scripts/approvalYears.py [hunk 9]
```diff
@@ -591,30 +594,30 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
     for entry in appInfo['table']:
         apps[entry[0]] = entry
 
-    for key in prods.keys():
-        if apps.has_key(key[0:6]):
+    for key in list(prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:842]: Same as above.

File: scripts/approvalYears.py [hunk 10]
```diff
@@ -591,30 +594,30 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
...
     for entry in submDates['table']:
-        if not subm.has_key(entry[0]):
+        if entry[0] not in subm:
             subm[entry[0]] = entry[5][0:10]
         elif subm[entry[0]] > entry[5][0:10]:
             subm[entry[0]] = entry[5][0:10]
 
-    for key in prods.keys():
-        if subm.has_key(key[0:6]):
+    for key in list(prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:864]: Same as above.

File: scripts/approvalYears.py [hunk 11]
```diff
@@ -732,10 +735,10 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
         appType = entry[5]
         prod = [appNo, product, '', '', '', appType, sponsor, date, dateRef]
         prods[appNo] = prod
-        if not UNII2prods.has_key(unii):
+        if unii not in UNII2prods:
             UNII2prods[unii] = []
         UNII2prods[unii].append(appNo)
-    for prod in prods.keys():
+    for prod in list(prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:986]: Same as above.

File: scripts/approvalYears.py [hunk 12]
```diff
@@ -765,7 +768,7 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
         if method == 'PREDICTED':
             appPredDate[appl] = date
 
-    for prod in prods.keys():
+    for prod in list(prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1016]: Same as above.

File: scripts/approvalYears.py [hunk 13]
```diff
@@ -806,14 +809,14 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
         match = 5
         if method == 'PREDICTED':
             early = [getTimeStamp(), 'Not available']
-            if not activeMoiety.has_key(unii): # active moiety was recently updated
-                for key in activeMoiety.keys():
+            if unii not in activeMoiety: # active moiety was recently updated
+                for key in list(activeMoiety.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1058]: Same as above.

File: scripts/approvalYears.py [hunk 14]
```diff
@@ -825,34 +828,34 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
...
 
                 for otherunii in activeMoiety[unii]:
-                    if UNII2prods.has_key(otherunii):
+                    if otherunii in UNII2prods:
                         for prod in UNII2prods[otherunii]:
-                            print otherunii, prod, prods[prod]
+                            print(otherunii, prod, prods[prod])
 
-                for prod in prods.keys():
+                for prod in list(prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1094]: Same as above.

File: scripts/approvalYears.py [hunk 15]
```diff
@@ -825,34 +828,34 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
...
                     if prod[0:6] == appl:
-                        print prod, prods[prod]
+                        print(prod, prods[prod])
 
-                for otherunii in UNII2prods.keys():
+                for otherunii in list(UNII2prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1098]: Same as above.

File: scripts/approvalYears.py [hunk 16]
```diff
@@ -872,7 +875,7 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
     #outfile2 =  maindir+"/temp/additionalWithdrawn-"+getTimeStamp()+".txt"
     #fperr = open(outfile2, 'w')
     #fperr.write(header)
-    for unii in activeMoiety.keys():
+    for unii in list(activeMoiety.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1123]: Same as above.

File: scripts/approvalYears.py [hunk 17]
```diff
@@ -923,10 +926,10 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
                         early[akey][-5] = entry[-5]
 
 
-        for key in early.keys():
+        for key in list(early.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1174]: Same as above.

File: scripts/approvalYears.py [hunk 18]
```diff
@@ -978,18 +981,18 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
             
     # write out all products data
     prod2UNIIs = dict()
-    for unii in UNII2prods.keys():
+    for unii in list(UNII2prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1229]: Same as above.

File: scripts/approvalYears.py [hunk 19]
```diff
@@ -978,18 +981,18 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
...
         for prod in UNII2prods[unii]:
-            if not prod2UNIIs.has_key(prod):
+            if prod not in prod2UNIIs:
                 prod2UNIIs[prod] = []
             prod2UNIIs[prod].append(unii)
 
     UNII2pt = dict()
 
-    for key in resolverCache.keys():
+    for key in list(resolverCache.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1237]: Same as above.

File: scripts/approvalYears.py [hunk 20]
```diff
@@ -978,18 +981,18 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
...
         UNII2pt[resolverCache[key]] = key
 
-    for key in uniiPT.keys():
+    for key in list(uniiPT.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1240]: Same as above.

File: scripts/approvalYears.py [hunk 21]
```diff
@@ -998,15 +1001,15 @@ def writeCBERBLAs(fpout, uniiPT, uniiALL):
     header = 'NDA\tProduct\tForm;Route\tStrength\tStatus\tAppl Type\tSponsor\tDate\tDate Ref\tUNIIs\tIngredients\n'
     fp.write(header)
     
-    for prod in prods.keys():
+    for prod in list(prods.keys()):
```
Comment (grishagin) [inline scripts/approvalYears.py:1249]: Same as above.

**Review by grishagin — CHANGES_REQUESTED** [review]
Synopsis: a typo and a likely code simplification suggestion in multiple places.
Just one more general thought (for the future): this thing has got to be completely refactored and modularized imo (so that someone else can understand what's going on here 😅).

File: scripts/approvalYears.py (hunk 1 above)
Comment (southalln) [reply scripts/approvalYears.py:291]: Variable name is correct ... "si" ... which is an awful name, but I was using it as "substance ingredient" in this case

File: scripts/approvalYears.py (hunk 2 above)
Comment (southalln) [reply scripts/approvalYears.py:210]: Python3 versus python2 issue. 2to3 automagically has made this change as .keys() no longer returns a list in python3

File: scripts/approvalYears.py (hunk 3 above)
Comment (southalln) [reply scripts/approvalYears.py:361]: Python3 versus python2 issue. 2to3 automagically has made this change as .keys() no longer returns a list in python3

File: scripts/approvalYears.py (hunk 4 above)
Comment (southalln) [reply scripts/approvalYears.py:385]: Python3 versus python2 issue. 2to3 automagically has made this change as .keys() no longer returns a list in python3

File: scripts/approvalYears.py (hunk 5 above)
Comment (southalln) [reply scripts/approvalYears.py:393]: Python3 versus python2 issue. 2to3 automagically has made this change as .keys() no longer returns a list in python3

File: scripts/approvalYears.py (hunk 17 above)
Comment (southalln) [reply scripts/approvalYears.py:1174]: Python3 versus python2 issue. 2to3 automagically has made this change as .keys() no longer returns a list in python3

File: scripts/approvalYears.py (hunk 18 above)
Comment (southalln) [reply scripts/approvalYears.py:1229]: Python3 versus python2 issue. 2to3 automagically has made this change as .keys() no longer returns a list in python3

File: scripts/approvalYears.py (hunk 19 above)
Comment (southalln) [reply scripts/approvalYears.py:1237]: Python3 versus python2 issue. 2to3 automagically has made this change as .keys() no longer returns a list in python3

File: scripts/approvalYears.py (hunk 20 above)
Comment (southalln) [reply scripts/approvalYears.py:1240]: Python3 versus python2 issue. 2to3 automagically has made this change as .keys() no longer returns a list in python3

File: scripts/approvalYears.py (hunk 21 above)
Comment (southalln) [reply scripts/approvalYears.py:1249]: Python3 versus python2 issue. 2to3 automagically has made this change as .keys() no longer returns a list in python3

File: scripts/approvalYears.py (hunk 1 above)
Comment (grishagin) [reply scripts/approvalYears.py:291]: 👍 I won't review the merge requests at night anymore... 😅

**Pull request merged**